File: ts/tests/router-bot.test.ts
import { jest } from '@jest/globals';
import type { IncomingMessage } from '../src/listener';

const mockSendMessage = jest.fn<(...args: unknown[]) => Promise<Record<string, string>>>()
  .mockResolvedValue({ id: 'msg-1' });
const mockSendCard = jest.fn<(...args: unknown[]) => Promise<Record<string, string>>>()
  .mockResolvedValue({ id: 'msg-1' });

jest.unstable_mockModule('../src/webex', () => ({
  sendMessage: mockSendMessage,
  sendCard: mockSendCard,
  getMessage: jest.fn(),
  getAttachmentAction: jest.fn(),
  extractCards: jest.fn(),
  _setMessagesUrl: jest.fn(),
}));

const { WgrokRouterBot } = await import('../src/router-bot');
import { loadCases } from './helpers';

interface RouterBotCases {
  config: { domains: string[] };
  routes: Record<string, string>;
  cases: Array<{
    name: string;
    sender: string;
    text: string;
    cards: unknown[];
    expect_send: boolean;
    expected_reply_to?: string;
    expected_reply_text?: string;
    expected_reply_card?: unknown;
    use_routes?: boolean;
  }>;
}

const CASES = loadCases<RouterBotCases>('router_bot_cases.json');

function fakeMsg(sender: string, text: string, html = '', roomId = '', roomType = ''): IncomingMessage {
  return { sender, text, html, roomId, roomType, msgId: 'test-msg-id', platform: 'webex', cards: [] };
}

describe('WgrokRouterBot', () => {
  beforeEach(() => {
    mockSendMessage.mockClear();
    mockSendCard.mockClear();
  });

  it.each(CASES.cases)('$name', async (tc) => {
    const routes = tc.use_routes ? CASES.routes : {};
    const bot = new WgrokRouterBot({
      webexToken: 'fake-token',
      domains: CASES.config.domains,
      debug: false,
      routes,
      platformTokens: { webex: ['fake-token'] },
      webhookPort: null,
      webhookSecret: null,
    });

    const msg = fakeMsg(tc.sender, tc.text);
    await bot.onMessageWithCards(msg, tc.cards);

    if (tc.expect_send) {
      const called = mockSendMessage.mock.calls.length > 0 || mockSendCard.mock.calls.length > 0;
      expect(called).toBe(true);

      if (tc.expected_reply_card) {
        expect(mockSendCard).toHaveBeenCalledTimes(1);
        const [, to, text] = mockSendCard.mock.calls[0] as [string, string, string];
        expect(to).toBe(tc.expected_reply_to);
        expect(text).toBe(tc.expected_reply_text);
      } else {
        expect(mockSendMessage).toHaveBeenCalledTimes(1);
        const [, to, text] = mockSendMessage.mock.calls[0] as [string, string, string];
        expect(to).toBe(tc.expected_reply_to);
        expect(text).toBe(tc.expected_reply_text);
      }
    } else {
      expect(mockSendMessage).not.toHaveBeenCalled();
      expect(mockSendCard).not.toHaveBeenCalled();
    }
  });

  describe('pause/resume', () => {
    it('pause command is recognized', async () => {
      const bot = new WgrokRouterBot({
        webexToken: 'fake-token',
        domains: ['example.com'],
        debug: false,
        routes: {},
        platformTokens: { webex: ['fake-token'] },
        webhookPort: null,
        webhookSecret: null,
      });

      const msg = fakeMsg('endpoint@example.com', './pause');
      await bot.onMessageWithCards(msg, []);

      expect(mockSendMessage).not.toHaveBeenCalled();
    });

    it('pause buffers messages sent back to paused endpoint', async () => {
      const bot = new WgrokRouterBot({
        webexToken: 'fake-token',
        domains: ['example.com'],
        debug: false,
        routes: {},
        platformTokens: { webex: ['fake-token'] },
        webhookPort: null,
        webhookSecret: null,
      });

      // Endpoint pauses (in Mode B, this is the same as sender)
      const pauseMsg = fakeMsg('endpoint@example.com', './pause');
      await bot.onMessageWithCards(pauseMsg, []);

      mockSendMessage.mockClear();

      // Send echo from endpoint to itself (Mode B echo back)
      const echoMsg = fakeMsg('endpoint@example.com', './echo:slug:from:-:test');
      await bot.onMessageWithCards(echoMsg, []);

      // Should buffer, not send
      expect(mockSendMessage).not.toHaveBeenCalled();
    });

    it('resume flushes buffered messages', async () => {
      const bot = new WgrokRouterBot({
        webexToken: 'fake-token',
        domains: ['example.com'],
        debug: false,
        routes: {},
        platformTokens: { webex: ['fake-token'] },
        webhookPort: null,
        webhookSecret: null,
      });

      // Endpoint pauses
      const pauseMsg = fakeMsg('endpoint@example.com', './pause');
      await bot.onMessageWithCards(pauseMsg, []);

      // Buffer some messages
      const echoMsg1 = fakeMsg('endpoint@example.com', './echo:slug:from:-:msg1');
      await bot.onMessageWithCards(echoMsg1, []);

      const echoMsg2 = fakeMsg('endpoint@example.com', './echo:slug:from:-:msg2');
      await bot.onMessageWithCards(echoMsg2, []);

      mockSendMessage.mockClear();

      // Resume
      const resumeMsg = fakeMsg('endpoint@example.com', './resume');
      await bot.onMessageWithCards(resumeMsg, []);

      // Should flush both buffered messages
      expect(mockSendMessage).toHaveBeenCalledTimes(2);
      const [, to1, text1] = mockSendMessage.mock.calls[0] as [string, string, string];
      expect(to1).toBe('endpoint@example.com');
      expect(text1).toBe('slug:from:-:msg1');

      const [, to2, text2] = mockSendMessage.mock.calls[1] as [string, string, string];
      expect(to2).toBe('endpoint@example.com');
      expect(text2).toBe('slug:from:-:msg2');
    });

    it('pause does not affect messages to other endpoints', async () => {
      const bot = new WgrokRouterBot({
        webexToken: 'fake-token',
        domains: ['example.com'],
        debug: false,
        routes: {},
        platformTokens: { webex: ['fake-token'] },
        webhookPort: null,
        webhookSecret: null,
      });

      // endpoint1 pauses
      const pauseMsg = fakeMsg('endpoint1@example.com', './pause');
      await bot.onMessageWithCards(pauseMsg, []);

      mockSendMessage.mockClear();

      // Send echo from endpoint2 (not paused)
      const echoMsg = fakeMsg('endpoint2@example.com', './echo:slug:from:-:test');
      await bot.onMessageWithCards(echoMsg, []);

      // Should send normally
      expect(mockSendMessage).toHaveBeenCalledTimes(1);
      const [, to] = mockSendMessage.mock.calls[0] as [string, string];
      expect(to).toBe('endpoint2@example.com');
    });
  });
});

File: rust/tests/router_bot_test.rs
use serde::Deserialize;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::fs;
use webex_message_handler::{DecryptedMessage, MercuryActivity};
use wiremock::matchers::{method, path};
use wiremock::{Mock, MockServer, ResponseTemplate};
use wgrok::{BotConfig, WgrokRouterBot, _set_messages_url};

#[derive(Deserialize)]
struct RouterBotCases {
    config: RouterBotConfig,
    routes: HashMap<String, String>,
    cases: Vec<RouterBotCase>,
}

#[derive(Deserialize)]
struct RouterBotConfig {
    domains: Vec<String>,
}

#[derive(Deserialize)]
struct RouterBotCase {
    name: String,
    sender: String,
    text: String,
    cards: Vec<Value>,
    expect_send: bool,
    #[allow(dead_code)]
    expected_reply_to: Option<String>,
    expected_reply_text: Option<String>,
    expected_reply_card: Option<Value>,
    #[serde(default)]
    use_routes: bool,
}

fn load_cases() -> RouterBotCases {
    let data = fs::read_to_string("../tests/router_bot_cases.json").expect("load router bot cases");
    serde_json::from_str(&data).expect("parse router bot cases")
}

fn fake_msg(sender: &str, text: &str) -> DecryptedMessage {
    DecryptedMessage {
        id: "test-msg-id".to_string(),
        room_id: "room-abc".to_string(),
        person_id: "person-123".to_string(),
        person_email: sender.to_string(),
        text: text.to_string(),
        html: None,
        created: "2024-01-01T00:00:00Z".to_string(),
        room_type: None,
        raw: MercuryActivity {
            id: String::new(),
            verb: String::new(),
            actor: Default::default(),
            object: Default::default(),
            target: Default::default(),
            published: String::new(),
            encryption_key_url: None,
        },
    }
}

#[tokio::test]
async fn test_router_bot_cases() {
    let cases = load_cases();

    for tc in &cases.cases {
        let server = MockServer::start().await;
        _set_messages_url(Some(format!("{}/messages", server.uri())));

        Mock::given(method("POST"))
            .and(path("/messages"))
            .respond_with(ResponseTemplate::new(200).set_body_json(json!({"id": "msg-1"})))
            .mount(&server)
            .await;

        let routes = if tc.use_routes {
            cases.routes.clone()
        } else {
            HashMap::new()
        };

        let bot = WgrokRouterBot::new(BotConfig {
            webex_token: "fake-token".to_string(),
            domains: cases.config.domains.clone(),
            debug: false,
            routes,
            platform_tokens: {
                let mut map = HashMap::new();
                map.insert("webex".to_string(), vec!["fake-token".to_string()]);
                map
            },
            webhook_port: None,
            webhook_secret: None,
        });

        let msg = fake_msg(&tc.sender, &tc.text);
        bot.on_message_with_cards(&msg, &tc.cards).await;

        let requests = server.received_requests().await.unwrap();

        if tc.expect_send {
            assert_eq!(
                requests.len(),
                1,
                "case {}: expected 1 request, got {}",
                tc.name,
                requests.len()
            );

            let body: Value = serde_json::from_slice(&requests[0].body).unwrap();
            // fake_msg always has room_id="room-abc", so bot routes via roomId
            assert_eq!(
                body["roomId"], "room-abc",
                "case {}: expected roomId routing",
                tc.name
            );
            // Verify toPersonEmail is absent when roomId is used
            assert!(
                body.get("toPersonEmail").is_none(),
                "case {}: toPersonEmail should not be set when roomId is present",
                tc.name
            );
            if let Some(ref expected_text) = tc.expected_reply_text {
                assert_eq!(
                    body["text"], *expected_text,
                    "case {}: reply_text mismatch",
                    tc.name
                );
            }
            if tc.expected_reply_card.is_some() {
                let has_attachments = body
                    .get("attachments")
                    .and_then(|a| a.as_array())
                    .map_or(false, |a| !a.is_empty());
                assert!(
                    has_attachments,
                    "case {}: expected card attachment",
                    tc.name
                );
            }
        } else {
            assert_eq!(
                requests.len(),
                0,
                "case {}: expected 0 requests, got {}",
                tc.name,
                requests.len()
            );
        }

        _set_messages_url(None);
    }
}
